### PR TITLE
Add new field and change unique index in result_scalar table; remove unique index in expected table

### PR DESCRIPTION
**Pull request by oiintam**
This is an updated version of what my past PR #29 and Christine's PR #28 tried to fix, which is the Pandokia import issue.

Basically a new field will be created in the database, and the rows of the field will contain a hash of `test_run+project+host+test_name+context`and this new field will serve as a new unique index for the result_scalar table. Since each subtest will have the a unique key id and if we would like to reimport the test results, we would like to have something other than key_id as unique index.
The hash is a string of length 32 so it will not exceed the limit of InnoDB.

Also, we would like to warn user if any test results is reimported when Pandokia is on verbose mode and to have a list of reimported test results to be written in a file for future reference.

In addition, I removed the unique index of the expected table due to the same reason happened in the old index of the result_scalar table. And there are some changes in the gen_expected script; since we remove unique index in expected table, we need to check if the row exists before inserting.

I have run two tests against our master branch with my test database and modified scripts:
https://ssb.stsci.edu/pandokia/c96.cgi?query=day_report.2&test_run=pandeia_master_2018-08-27-10%3A48%3A41
https://ssb.stsci.edu/pandokia/c96.cgi?query=day_report.2&test_run=pandeia_master_latest

### CODE REVIEW

**Comment by vglaidler** [comment]
Can you please explain in a bit more detail how you tested this? The JWST ETC test runs in and of themselves don't demonstrate that the Pandokia changes are correct.

For example, identifying the various possible cases along with what the behavior should be, and showing that that is the behavior achieved. (Hm, that sounds a bit vague - let's chat in person if you'd like more detail.)

**Comment by oiintam** [comment]
Manual test cases:
- reimporting test results -
I manually reimport some test results, for example,
import files `PDK_LOG.pandeia_master_2018-08-27-10:48:41.final.0` and `PDK_LOG.pandeia_master_2018-08-27-10:48:41.final.1` and the result is following:
```
% pdk import PDK_LOG.pandeia_master_2018-08-27-10:48:41.final*
FILE: PDK_LOG.pandeia_master_2018-08-27-10:48:41.final.0
2 records


FILE: PDK_LOG.pandeia_master_2018-08-27-10:48:41.final.1
WARNING: About to reimport the test results, please take a look at the reimport_tests.txt file after the import is done
WARNING: About to reimport the test results, please take a look at the reimport_tests.txt file after the import is done
WARNING: About to reimport the test results, please take a look at the reimport_tests.txt file after the import is done
3 records


There are 3 reimport tests during this run
The reimport tests file is called : reimport_tests_PDK_LOG.pandeia_master_2018-08-27-10:48:41.final.1.txt
%
```

**Comment by cslocum** [comment]
@oiintam, please update the description (no new field for the expected table).

**Comment by cslocum** [comment]
Is the output above produced in verbose mode?

**Comment by cslocum** [comment]
Joe - I think that would prohibit us from re-importing the same test output files, which is explicitly something we want to be able to do.  Or am I missing your point?

**Comment by jhunkeler** [comment]
Ha, sorry! A little context goes a long way :grinning: 

The scenario played out in my head like this... 

1) Human1 performs a test run
```
PDK_TESTRUN="one_off_but_still_important"  # No timestamp
PDK_CONTEXT="default"
PDK_[etc, etc]=...

$ pdkrun -r some/path/
$ pdk import ${PDK_LOGS}*
```
2) Human2 uses the same set of `PDK_*` environment variables a few weeks later and performs the same test run.

3) Human1 is disgruntled because Human2 accidentally destroyed their results.

----
It has been a while since I've really dug into the code so maybe I'm way off in my assumptions here. If so feel free to ignore this and carry on :+1:

**Comment by cslocum** [comment]
> Human1 is disgruntled because Human2 accidentally destroyed their results.

We want Human1 to be disgruntled :)  This is the exact scenario we want to allow.  The original code was designed to allow this - we aren't breaking any paradigms here.

**Comment by cdsontag** [comment]
I could go either way on this (i.e. I see Joe's point too), but I'll note that Human1 and Human2 can avoid the situation simply by adding a datestamp to their test name.

**Comment by cslocum** [comment]
> Human1 and Human2 can avoid the situation simply by adding a datestamp to their test name.

We almost always do this.

**Comment by oiintam** [comment]
Since the PR changes both the code base and the database structure, please keep in mind that in case of a new release/installation of Pandokia, the existing database 'dystopia' will need to have its current structure modified in order to accompany these script changes. Thank you. :)

File: pandokia/import_data.py [hunk 1]
```diff
@@ -278,9 +284,26 @@ def try_insert(self, db, key_id):
                  self.attn,
                  self.test_runner,
                  okf]
+        
+        hash_str = self.test_run+self.host+self.project+self.test_name+self.context
+
+        # return a string of length 32
+        h = hashlib.md5(hash_str.encode())
```
Comment (cdsontag) [inline pandokia/import_data.py]: Did you verify these lines are the same in py2 and py3?  If not, you can simply play with md5 calls in two different interpreter windows in your terminal and make sure they take/return the types you are expecting that they do.

Comment (oiintam) [reply pandokia/import_data.py]: Yes, I have tested the hashlib functions in both python 2 and 3, they both work as expected.

File: pandokia/import_data.py [hunk 2]
```diff
@@ -476,7 +501,15 @@ def run(argv, hack_callback=None):
                     print('Skipped: {}'.format(rx.test_name))
                 duplicate_count += 1
 
-            pdk_db.commit()
+            pdk_db.commit() 
+        
+        if reimport_count != 0:
+            print("There are {:d} reimport tests during this run".format(reimport_count))
```
Comment (vglaidler) [inline pandokia/import_data.py]: Can we please change this message to "{:d} tests were re-imported during this run"
and the subsequent one to "See details in %s" ?

**Comment by vglaidler** [comment]
The only additional comment I have is that there was a LOT of discussion while working this PR about whether and when a unique index was needed and when it caused problems. It would be good to add a paragraph or so discussion of this in the schema in the appropriate places. 

I suggest replacing the terse "we only need one entry" with this more extensive explanation.

File: pandokia/import_data.py [hunk 3]
```diff
@@ -278,9 +284,26 @@ def try_insert(self, db, key_id):
...
+        parm += [self.hash]
+
+        a = db.execute("select status from result_scalar where test_hash = :1", (self.hash,))
+        y = a.fetchone()
+        if y is not None:
+            db.execute("delete from result_scalar where test_hash = :1", (self.hash,))
+            reimport_count += 1
+            reimport_parm[reimport_count] = parm
+            if not quiet:
+                print("WARNING: About to reimport the test results, please take a look at the reimport_tests.txt file after the import is done")
```
[the remark this answers is not in the document]
Comment (oiintam) [reply pandokia/import_data.py]: It is about to reimport a single test result.

File: pandokia/import_data.py (hunk 2 above)
Comment (oiintam) [reply pandokia/import_data.py]: Will do.

File: pandokia/import_data.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (oiintam) [reply pandokia/import_data.py]: I see a mistake on grammar, I will fix that.

**Comment by oiintam** [comment]
I have made changes that @vglaidler had requested. @cslocum mentioned that we should merge this before our sprint starts, so I will merge this today if no one objects. Thanks.

**Comment by cdsontag** [comment]
@cslocum and I think this is ready to go in and all concerns have been covered.
@jhunkeler it looks like you are OK with this too.

**Comment by cdsontag** [comment]
We all, including @oiintam think this is merge-able.  Merging.

**Pull request merged**